### PR TITLE
SQL Edge deployment using Azure IoT hub

### PR DESCRIPTION
**Pull request by alanrenmsft**
1. New deployment scenario: SQL Edge through Azure IoT hub
1. improve the azure_accounts field type: add sign-in and refresh feature for accounts drop down and also allow creation of new Resource Group
1. updated the linked account dialog logic so that the command will only return when the dialog is closed.
Screen Shot 2020-07-03 at 4 32 27 PM

### CODE REVIEW

File: extensions/asde-deployment/package.nls.json [hunk 1]
```diff
@@ -26,5 +26,18 @@
 	"remote-info-section-title": "Target machine information",
 	"edge-remote-target-field": "Name or IP address",
 	"edge-remote-username-field": "Username",
-	"edge-remote-password-field": "Password"
+	"edge-remote-password-field": "Password",
+	"sql-edge-azure-display-name": "An Azure VM via Azure IoT Hub",
+	"sql-edge-azure-notebook": "./notebooks/edge/deploy-sql-edge-azure.ipynb",
```
Comment (Charles-Gagnon) [inline extensions/asde-deployment/package.nls.json:31]: Why are these paths in the localized file? I wouldn't expect these to ever be actually localized right?

File: extensions/resource-deployment/src/ui/modelViewUtils.ts [hunk 2]
```diff
@@ -602,18 +608,25 @@ function processCheckboxField(context: FieldContext): void {
  * @param context The context to use to create the field
  */
 function processFilePickerField(context: FieldContext): FilePickerInputs {
+	const inputWidth = parseInt(context.fieldInfo.inputWidth!);
+	if (inputWidth === NaN) {
+		// this is a dev time only error
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/modelViewUtils.ts:613]: This isn't really true, could happen in a stable build too just as easily if someone didn't test thoroughly.

Anyways I would suggest you look into getting an actual JSON schema set up for the resource deployment contribution as well so that these mistakes are easier to avoid. Right now we have a mix of straight int values and '##px' string values for inputWidth values so I think it'd be best to just be completely consistent here and switch everything over to int if you're going to be doing this type of logic. Or you could add logic to check for px and trim it off - whatever your preference is I guess.

File: extensions/resource-deployment/src/ui/modelViewUtils.ts [hunk 3]
```diff
@@ -622,7 +635,7 @@ function processFilePickerField(context: FieldContext): FilePickerInputs {
 			defaultUri: vscode.Uri.file(path.dirname(input.value || os_homedir())),
 			openLabel: loc.select,
 			filters: {
-				'Config Files': ['*'],
+				'File': ['*'],
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/modelViewUtils.ts:638]: Localize?

File: extensions/resource-deployment/src/ui/modelViewUtils.ts [hunk 4]
```diff
@@ -731,31 +745,66 @@ async function processAzureAccountField(context: AzureAccountFieldContext): Prom
 	context.fieldInfo.subFields = [];
 	const accountValueToAccountMap = new Map<string, azdata.Account>();
 	const subscriptionValueToSubscriptionMap = new Map<string, azureResource.AzureResourceSubscription>();
-	const accountDropdown = createAzureAccountDropdown(context);
+	const accountComponents = createAzureAccountDropdown(context);
+	const accountDropdown = accountComponents.accountDropdown;
 	const subscriptionDropdown = createAzureSubscriptionDropdown(context, subscriptionValueToSubscriptionMap);
 	const resourceGroupDropdown = createAzureResourceGroupsDropdown(context, accountDropdown, accountValueToAccountMap, subscriptionDropdown, subscriptionValueToSubscriptionMap);
+	if (context.fieldInfo.allowNewResourceGroup) {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/modelViewUtils.ts:752]: Did you run this by design? I really don't like that we have two fields for RG - what changes when you check the checkbox? It seems like we should just have the single editable dropdown and just make it clear that a new RG will be created if the name doesn't exist (maybe have a conditional text field that only shows up if the currently entered name isn't in the list of RGs)

File: src/sql/workbench/services/accountManagement/browser/accountManagementService.ts [hunk 5]
```diff
@@ -297,9 +297,10 @@ export class AccountManagementService implements IAccountManagementService {
 	// UI METHODS //////////////////////////////////////////////////////////
 	/**
 	 * Opens the account list dialog
-	 * @return Promise that finishes when the account list dialog opens
+	 * @param waitForDialogClose whether to wait for the dialog to close
+	 * @return Promise that finishes when the account list dialog opens or closes based on the value of waitForDialogClose parameter
 	 */
-	public openAccountListDialog(): Thenable<void> {
+	public openAccountListDialog(waitForDialogClose: boolean): Thenable<void> {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/accountManagement/browser/accountManagementService.ts:303]: Hm, I guess this is fine but I'm not a huge fan of using parameters for this kind of thing (changing what the return value means).

What about just returning a nested Thenable, with the internal one being one that resolves when the dialog is closed?

`Thenable<Thenable<void>>`

File: extensions/resource-deployment/src/ui/modelViewUtils.ts (hunk 4 above)
Comment (alanrenmsft) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts:752]: Screen Shot 2020-07-06 at 1 57 18 PM

File: extensions/asde-deployment/package.nls.json (hunk 1 above)
Comment (alanrenmsft) [reply extensions/asde-deployment/package.nls.json:31]: i guess you are right, i will move this to package.json to avoid confusing the localization team.

File: extensions/resource-deployment/src/ui/modelViewUtils.ts (hunk 4 above)
Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts:752]: Alrighty, my opinion is that the checkbox should control what controls are enabled/disabled at least then. So when unchecked dropdown is enabled and text box is disabled. And vice versa for checked.

Comment (alanrenmsft) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts:752]: yep, it is already implemented that way.

File: extensions/resource-deployment/src/ui/modelViewUtils.ts (hunk 3 above)
Comment (alanrenmsft) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts:638]: I will expose this as a property of file_picker field, localization will be handled by extension.

Comment (alanrenmsft) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts]: **filter not specified**
Screen Shot 2020-07-06 at 7 43 31 PM

**filter specified**
Screen Shot 2020-07-06 at 7 43 51 PM

File: src/sql/workbench/services/accountManagement/browser/accountManagementService.ts (hunk 5 above)
Comment (alanrenmsft) [reply src/sql/workbench/services/accountManagement/browser/accountManagementService.ts]: I introduced this parameter to simplify the unit tests, but now i found an easy way to mock the account dialog as well, so this is no longer needed, we don't really have a scenario now that needs this to be conditional, so change it to make it always wait for the dialog close.

File: src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts [hunk 6]
```diff
@@ -421,13 +430,25 @@ suite('Account Management Service Tests:', () => {
...
+		let mockAccountDialogCloseEvent = new Emitter<void>();
+		mockAccountDialog['onCloseEvent'] = mockAccountDialogCloseEvent.event;
+		setTimeout(() => {
+			mockAccountDialogCloseEvent.fire();
+		}, 1000);
 		state.instantiationService.setup(x => x.createInstance(TypeMoq.It.isValue(AccountDialogController)))
 			.returns(() => mockDialogController.object);
 
 		// If: I open the account dialog for a second time
 		return state.accountManagementService.openAccountListDialog()
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts:445]: So this test is actually broken as it is - since it doesn't use a done callback it will just completely immediately once the promise is returned instead of waiting for the verifies to be called. 

It'd be nice to fix this now - but if we do so that would be in a separate PR I'd say.

File: extensions/resource-deployment/src/ui/modelViewUtils.ts (hunk 2 above)
Comment (alanrenmsft) [reply extensions/resource-deployment/src/ui/modelViewUtils.ts:613]: I am not planning to take a task of that size right now, so i am just going to remove the px from the width string for now.

File: src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts (hunk 6 above)
Comment (alanrenmsft) [reply src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts:445]: mocha works with Promise, if you return a promise, it will wait for the promise to be fulfilled:
https://mochajs.org/#working-with-promises

Comment (Charles-Gagnon) [reply src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts:445]: Oh right. Although I'm not sure what the intention is with firing the dialog close event. And why is that on a timer? That seems like a possibility of a race condition if you're relying on that event to trigger some state to change because of that...

Comment (alanrenmsft) [reply src/sql/workbench/services/accountManagement/test/browser/accountManagementService.test.ts:445]: since we are waiting for the account dialog to be closed now, we have to mock the dialog close event, otherwise the test will fail with timeout error. I am trying to fire the event after the controller.openAccountDialog() is called, below is the accountManagementService's implementation, all of the objects involved are mocked and the execution shouldn't take much time, 1 sec is more than enough for now without needing to introduce more complicated mechanism.
`
	public openAccountListDialog(): Thenable<void> {
		let self = this;

		return new Promise((resolve, reject) => {
			try {
				// If the account list dialog hasn't been defined, create a new one
				if (!self._accountDialogController) {
					self._accountDialogController = self._instantiationService.createInstance(AccountDialogController);
				}
				self._accountDialogController.openAccountDialog();
				self._accountDialogController.accountDialog.onCloseEvent(resolve);
			} catch (e) {
				reject(e);
			}
		});
	}
`
